raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -2803,21 +2803,28 @@ public void testDescribeQuorumWithOnlyStaticVoters(boolean withKip853Rpc) throws
     @CsvSource({ "true, true", "true, false", "false, false" })
     public void testDescribeQuorumWithFollowers(boolean withKip853Rpc, boolean withBootstrapSnapshot) throws Exception {
         int localId = randomReplicaId();
-        ReplicaKey local = replicaKey(localId, true);
-        Uuid localDirectoryId = local.directoryId().get();
-        ReplicaKey follower1 = replicaKey(localId + 1, true);
-        Uuid followerDirectoryId1 = follower1.directoryId().get();
-        ReplicaKey follower2 = replicaKey(localId + 2, false);
-        Set<Integer> voters = Utils.mkSet(localId, follower1.id(), follower2.id());
-        VoterSet voterSet = VoterSetTest.voterSet(Stream.of(local, follower1, follower2));
+        int followerId1 = localId + 1;
+        int followerId2 = localId + 2;
+        ReplicaKey local = replicaKey(localId, withBootstrapSnapshot);
+        // local directory id must exist
+        Uuid localDirectoryId = local.directoryId().orElse(Uuid.randomUuid());
+        ReplicaKey bootstrapFollower1 = replicaKey(followerId1, withBootstrapSnapshot);
+        // if withBootstrapSnapshot is false, directory ids are still needed by the static voter set
+        Uuid followerDirectoryId1 = bootstrapFollower1.directoryId().orElse(withKip853Rpc ? Uuid.randomUuid() : ReplicaKey.NO_DIRECTORY_ID);
+        ReplicaKey follower1 = ReplicaKey.of(followerId1, followerDirectoryId1);
+        ReplicaKey bootstrapFollower2 = replicaKey(followerId2, withBootstrapSnapshot);
+        Uuid followerDirectoryId2 = bootstrapFollower2.directoryId().orElse(withKip853Rpc ? Uuid.randomUuid() : ReplicaKey.NO_DIRECTORY_ID);
+        ReplicaKey follower2 = ReplicaKey.of(followerId2, followerDirectoryId2);
+        VoterSet bootstrapVoterSet = VoterSetTest.voterSet(Stream.of(local, bootstrapFollower1, bootstrapFollower2));
+        VoterSet staticVoterSet = VoterSetTest.voterSet(Stream.of(local, follower1, follower2));
```

Comment:
You can move this to the `if (withBootstrapSnapshot) ... else ...` state and avoid creating too many local variable that are not used through out the method.

This comment applies to a few places.